ipld/resolver/src/behaviour/discovery.rs

```diff
@@ -0,0 +1,292 @@
+            .user_defined
+            .iter()
+            .filter(|(p, _)| p == peer_id)
+            .map(|(_, a)| a.clone())
+            .collect::<Vec<_>>();
+
+        addrs.extend(self.inner.addresses_of_peer(peer_id));
+        addrs
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
+        match &event {
+            FromSwarm::ConnectionEstablished(e) => {
+                self.num_connections += 1;
+                if e.other_established == 0 {
+                    let addrs = self.addresses_of_peer(&e.peer_id);
+                    self.outbox
+                        .push_back(DiscoveryEvent::Connected(e.peer_id, addrs));
+                }
+            }
+            FromSwarm::ConnectionClosed(e) => {
+                self.num_connections -= 1;
+                if e.remaining_established == 0 {
+                    let addrs = self.addresses_of_peer(&e.peer_id);
+                    self.outbox
+                        .push_back(DiscoveryEvent::Disconnected(e.peer_id, addrs));
+                }
+            }
+            _ => {}
+        };
+        self.inner.on_swarm_event(event)
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        peer_id: PeerId,
+        connection_id: ConnectionId,
+        event: <<Self::ConnectionHandler as IntoConnectionHandler>::Handler as ConnectionHandler>::OutEvent,
+    ) {
+        self.inner
+            .on_connection_handler_event(peer_id, connection_id, event)
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+        params: &mut impl PollParameters,
+    ) -> std::task::Poll<NetworkBehaviourAction<Self::OutEvent, Self::ConnectionHandler>> {
+        // Emit own events first.
+        if let Some(ev) = self.outbox.pop_front() {
+            return Poll::Ready(NetworkBehaviourAction::GenerateEvent(ev));
+        }
+
+        // Trigger periodic queries.
+        if self.lookup_interval.poll_tick(cx).is_ready() {
+            if self.num_connections < self.max_connections {
+                if let Some(k) = self.inner.as_mut() {
+                    let random_peer_id = PeerId::random();
+                    k.get_closest_peers(random_peer_id);
+                }
+            }
+
+            // Schedule the next random query with exponentially increasing delay, capped at 60 seconds.
+            self.lookup_interval = tokio::time::interval(cmp::min(
+                self.lookup_interval.period() * 2,
+                Duration::from_secs(60),
+            ));
+            // we need to reset the interval, otherwise the next tick completes immediately.
+            self.lookup_interval.reset();
+        }
+
+        // Poll Kademlia.
+        while let Poll::Ready(ev) = self.inner.poll(cx, params) {
+            match ev {
+                // Not propagating Kademlia specific events, just the ones meant for the Swarm.
+                NetworkBehaviourAction::GenerateEvent(_out) => {
```

Comment:
Good question. Forest ignored them so kept that, but some of the context was lost:

```rust
                NetworkBehaviourAction::GenerateEvent(ev) => match ev {
                    // Adding to Kademlia buckets is automatic with our config,
                    // no need to do manually.
                    KademliaEvent::RoutingUpdated { .. } => {}
                    KademliaEvent::RoutablePeer { .. } => {}
                    KademliaEvent::PendingRoutablePeer { .. } => {
                        // Intentionally ignore
                    }
                    other => {
                        debug!("Libp2p => Unhandled Kademlia event: {:?}", other)
                    }
                },
```

For the record, Kademlia has these events:

<details>
<summary>KademliaEvent</summary>
```rust
pub enum KademliaEvent {
    /// An inbound request has been received and handled.
    //
    // Note on the difference between 'request' and 'query': A request is a
    // single request-response style exchange with a single remote peer. A query
    // is made of multiple requests across multiple remote peers.
    InboundRequest { request: InboundRequest },

    /// An outbound query has made progress.
    OutboundQueryProgressed {
        /// The ID of the query that finished.
        id: QueryId,
        /// The intermediate result of the query.
        result: QueryResult,
        /// Execution statistics from the query.
        stats: QueryStats,
        /// Indicates which event this is, if therer are multiple responses for a single query.
        step: ProgressStep,
    },

    /// The routing table has been updated with a new peer and / or
    /// address, thereby possibly evicting another peer.
    RoutingUpdated {
        /// The ID of the peer that was added or updated.
        peer: PeerId,
        /// Whether this is a new peer and was thus just added to the routing
        /// table, or whether it is an existing peer who's addresses changed.
        is_new_peer: bool,
        /// The full list of known addresses of `peer`.
        addresses: Addresses,
        /// Returns the minimum inclusive and maximum inclusive [`Distance`] for
        /// the bucket of the peer.
        bucket_range: (Distance, Distance),
        /// The ID of the peer that was evicted from the routing table to make
        /// room for the new peer, if any.
        old_peer: Option<PeerId>,
    },

    /// A peer has connected for whom no listen address is known.
    ///
    /// If the peer is to be added to the routing table, a known
    /// listen address for the peer must be provided via [`Kademlia::add_address`].
    UnroutablePeer { peer: PeerId },

    /// A connection to a peer has been established for whom a listen address
    /// is known but the peer has not been added to the routing table either
    /// because [`KademliaBucketInserts::Manual`] is configured or because
    /// the corresponding bucket is full.
    ///
    /// If the peer is to be included in the routing table, it must
    /// must be explicitly added via [`Kademlia::add_address`], possibly after
    /// removing another peer.
    ///
    /// See [`Kademlia::kbucket`] for insight into the contents of
    /// the k-bucket of `peer`.
    RoutablePeer { peer: PeerId, address: Multiaddr },

    /// A connection to a peer has been established for whom a listen address
    /// is known but the peer is only pending insertion into the routing table
    /// if the least-recently disconnected peer is unresponsive, i.e. the peer
    /// may not make it into the routing table.
    ///
    /// If the peer is to be unconditionally included in the routing table,
    /// it should be explicitly added via [`Kademlia::add_address`] after
    /// removing another peer.
    ///
    /// See [`Kademlia::kbucket`] for insight into the contents of
    /// the k-bucket of `peer`.
    PendingRoutablePeer { peer: PeerId, address: Multiaddr },
}

```
</details>

I'll add more comments. From what I see only `UnroutablePeer` would be surprising.

Reply:
Added comments and a `debug!` for the `UnroutablePeer` specifically.